File: sample-chat-obj-c/sample-chat/Components/DialogsScreen/DialogsViewController.m
//
//  DialogsViewController.m
//  sample-chat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "DialogsViewController.h"
#import "DialogsSelectionViewController.h"
#import "CreateNewDialogViewController.h"
#import "UIViewController+InfoScreen.h"
#import "ImageCache.h"
#import "Profile.h"
#import "UINavigationController+Appearance.h"
#import "QBChatMessage+Chat.h"
#import "UIViewController+Alert.h"
#import "ConnectionModule.h"
#import "AuthModule.h"
#import "NotificationsProvider.h"
#import "SplashScreenViewController.h"
#import "NSError+Chat.h"

@interface DialogsViewController () <ChatManagerDelegate, QBChatDelegate, AuthModuleDelegate, ConnectionModuleDelegate>
//MARK: - Properties
@property (nonatomic, strong) Profile *profile;
@property (strong, nonatomic) ConnectionModule *connection;
@property (strong, nonatomic) AuthModule *authModule;
@property (strong, nonatomic) SplashScreenViewController *splashVC;
@end

NSString *const kStillConnection = @"Still in connecting state, please wait";
NSString *const kReconnection = @"Reconnecting state, please wait";

@implementation DialogsViewController
- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.authModule = [[AuthModule alloc] init];
    self.authModule.delegate = self;
    self.connection = [[ConnectionModule alloc] init];
    self.connection.delegate = self;
    if (!self.connection.established) {
        [self showSplashScreen];
    }
    [self.connection activateAutomaticMode];
    
    self.chatManager.delegate = self;
    self.profile = [[Profile alloc] init];
    [QBChat.instance addDelegate:self];
    
    UILongPressGestureRecognizer *tapGestureDelete = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(tapEdit:)];
    tapGestureDelete.minimumPressDuration = 0.5f;
    tapGestureDelete.delaysTouchesBegan = YES;
    [self.tableView addGestureRecognizer:tapGestureDelete];
}

- (void)setupDialogs {
    [self.refreshControl beginRefreshing];
    [self.chatManager updateStorage];
}

- (void)setupViews {
    self.navigationItem.rightBarButtonItems = @[];
    self.navigationItem.leftBarButtonItems = @[];
    UIBarButtonItem *exitButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"exit"]
                                                                       style:UIBarButtonItemStylePlain
                                                                      target:self
                                                                      action:@selector(logoutButtonPressed:)];
    exitButtonItem.tintColor = UIColor.whiteColor;
    
    UIBarButtonItem *emptyButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"icon-info"]
                                                                        style:UIBarButtonItemStylePlain
                                                                       target:self
                                                                       action:nil];
    emptyButtonItem.tintColor = UIColor.clearColor;
    emptyButtonItem.enabled = NO;
    self.navigationItem.leftBarButtonItems = @[exitButtonItem, emptyButtonItem];
    UIBarButtonItem *usersButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"add"]
                                                                        style:UIBarButtonItemStylePlain
                                                                       target:self
                                                                       action:@selector(didTapNewChat:)];
    self.navigationItem.rightBarButtonItem = usersButtonItem;
    usersButtonItem.tintColor = UIColor.whiteColor;
    [self addInfoButton];
}

- (void)setupNavigationTitle {
    self.navigationItem.title = @"Chats";
}

#pragma mark - Actions

- (IBAction)refreshDialogs:(UIRefreshControl *)sender {
    [self.chatManager updateStorage];
}

- (void)tapEdit:(UILongPressGestureRecognizer *)gestureReconizer {
    if (gestureReconizer.state == UIGestureRecognizerStateEnded) {
        return;
    }
    gestureReconizer.state = UIGestureRecognizerStateEnded;
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Dialogs" bundle:nil];
    DialogsSelectionViewController *deleteVC = [storyboard instantiateViewControllerWithIdentifier:@"DialogsSelectionVC"];
    deleteVC.action = ChatActionDelete;
    [self.navigationController pushViewController:deleteVC animated:NO];
}

- (void)didTapNewChat:(UIButton *)sender {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Dialogs" bundle:nil];
    CreateNewDialogViewController *createNewDialogVC = [storyboard instantiateViewControllerWithIdentifier:@"CreateNewDialogViewController"];
    [self.navigationController pushViewController:createNewDialogVC animated:YES ];
}

#pragma mark Logout
- (void)logoutButtonPressed:(UIButton *)sender {
    [self logout];
}

- (void)logout {
    if (!self.connection.established) {
        [self showNoInternetAlertWithHandler:nil];
        return;
    }
    [self.progressView start];
    __weak __typeof(self)weakSelf = self;
    [NotificationsProvider deleteLastSubscriptionWithCompletion:^{
        [weakSelf.connection breakConnectionWithCompletion:^{
            [weakSelf.authModule logout];
        }];
    }];
}

- (void)openChatScreenWithDialogID:(NSString *)dialogID {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Chat" bundle:nil];
    ChatViewController *chatVC = [storyboard instantiateViewControllerWithIdentifier:@"ChatViewController"];
    chatVC.dialogID = dialogID;
    [self.navigationController pushViewController:chatVC animated:NO];
}

- (void)configureCell:(DialogCell *)cell
          forIndexPath:(NSIndexPath *)indexPath {
    self.tableView.allowsMultipleSelection = NO;
    cell.checkBoxImageView.hidden = YES;
    cell.checkBoxView.hidden = YES;
    cell.lastMessageDateLabel.hidden = NO;
    cell.contentView.backgroundColor = UIColor.clearColor;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    QBChatDialog *dialog = self.dialogs[indexPath.row];
    [self openChatScreenWithDialogID:dialog.ID];
}

- (void)showSplashScreen {
    if (self.splashVC) {
        return;
    }
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Authorization" bundle:nil];
    self.splashVC = [storyboard instantiateViewControllerWithIdentifier:@"SplashScreenViewController"];
    self.splashVC.modalPresentationStyle = UIModalPresentationOverCurrentContext;
    [self presentViewController:self.splashVC animated:NO completion:nil];
}

- (void)hideSplashScreen {
    if (!self.splashVC) {
        return;
    }
    [self.splashVC dismissViewControllerAnimated:NO completion:^{
        self.splashVC = nil;
    }];
}

#pragma mark QBChatDelegate
- (void)chatDidReceiveMessage:(QBChatMessage *)message {
    [self.chatManager updateDialogWith:message.dialogID withMessage:message];
}

- (void)chatRoomDidReceiveMessage:(QBChatMessage *)message fromDialogID:(NSString *)dialogID {
    [self.chatManager updateDialogWith:message.dialogID withMessage:message];
}

- (void)chatDidReceiveSystemMessage:(QBChatMessage *)message {
    if ([self.chatManager.storage dialogWithID:message.dialogID]) {
        return;
    }
    NSString *dialogID = message.dialogID ? : message.customParameters[@"dialog_Id"];
    if (dialogID) {
        [self.chatManager updateDialogWith:dialogID withMessage:message];
    }
}

#pragma mark Chat Manager Delegate
- (void)chatManager:(ChatManager *)chatManager didUpdateStorage:(NSString *)message {
    [self reloadContent];
}

- (void)chatManager:(ChatManager *)chatManager didUpdateChatDialog:(QBChatDialog *)chatDialog {
    [self reloadContent];
}

- (void)chatManager:(ChatManager *)chatManager didFailUpdateStorage:(NSString *)message {
    [self.refreshControl endRefreshing];
    [self showAnimatedAlertWithTitle:nil message:message];
}

#pragma mark - AuthModuleDelegate
- (void)authModule:(AuthModule *)authModule didLoginUser:(QBUUser *)user {
    [Profile synchronizeUser:user];
    [self.connection establish];
}

- (void)authModuleDidLogout:(AuthModule *)authModule {
    [self.connection deactivateAutomaticMode];
    [self.navigationController popToRootViewControllerAnimated:NO];
    [Profile clear];
    [self.chatManager.storage clear];
    [ImageCache.instance clearCache];
    [NSUserDefaults.standardUserDefaults removeObjectForKey:kToken];
    [self.progressView stop];
    if (self.onSignOut) {
        self.onSignOut();
    }
}

- (void)authModule:(AuthModule *)authModule didReceivedError:(NSError *)error {
    [self showUnAuthorizeAlert:error.localizedDescription logoutAction:^(UIAlertAction * _Nonnull action) {
        [self logout];
    } tryAgainAction:^(UIAlertAction * _Nonnull action) {
        [authModule loginWithFullName:self.profile.fullName login:self.profile.login];
    }];
}

#pragma mark - ConnectionModuleDelegate
- (void)connectionModuleWillConnect:(ConnectionModule *)connectionModule {
    [self showAnimatedAlertWithTitle:nil message:kStillConnection];
}

- (void)connectionModuleDidConnect:(ConnectionModule *)connectionModule {
    [self setupDialogs];
    [self hideAlertView];
    [self hideSplashScreen];
}

- (void)connectionModuleDidNotConnect:(ConnectionModule *)connectionModule withError:(NSError*)error {
    [self.refreshControl endRefreshing];
    [self hideSplashScreen];
    if (error.isNetworkError) {
        [self showNoInternetAlertWithHandler:nil];
        return;
    }
    [self showAlertWithTitle:nil message:error.localizedDescription handler:nil];
}

- (void)connectionModuleWillReconnect:(ConnectionModule *)connectionModule {
    [self showAnimatedAlertWithTitle:nil message:kReconnection];
}

- (void)connectionModuleDidReconnect:(ConnectionModule *)connectionModule {
    [self setupDialogs];
    [self hideAlertView];
}

- (void)connectionModuleTokenHasExpired:(ConnectionModule *)connectionModule {
    [self showSplashScreen];
    [self.refreshControl endRefreshing];
    Profile *profile = [[Profile alloc] init];
    [self.authModule loginWithFullName:profile.fullName login:profile.login];
}

@end

File: sample-chat-swift/sample-chat-swift/Components/DialogsScreen/DialogsViewController.swift
//
//  DialogsViewController.swift
//  sample-chat-swift
//
//  Created by Injoit on 9/30/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit
import Quickblox

class DialogsViewController: DialogListViewController {
    //MARK: - Properties
    var onSignOut: (() -> Void)?
    var splashVC: SplashScreenViewController!
    private let authModule = AuthModule()
    private let connection = ConnectionModule()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        connection.delegate = self
        if connection.established == false {
            showSplashScreen()
        }
        connection.activateAutomaticMode()
        authModule.delegate = self
        QBChat.instance.addDelegate(self)
        chatManager.delegate = self
        
        let tapGestureDelete = UILongPressGestureRecognizer(target: self, action: #selector(didTapEdit(_:)))
        tapGestureDelete.minimumPressDuration = 0.5
        tapGestureDelete.delaysTouchesBegan = true
        tableView.addGestureRecognizer(tapGestureDelete)
    }
    
    //MARK: - Setup
    override func setupNavigationBar() {
        navigationItem.rightBarButtonItems = []
        navigationItem.leftBarButtonItems = []
        let exitButtonItem = UIBarButtonItem(image: UIImage(named: "exit"),
                                             style: .plain,
                                             target: self,
                                             action: #selector(didTapLogout))
        exitButtonItem.tintColor = .white
        
        let emptyButtonItem = UIBarButtonItem(image: UIImage(named: "icon-info"),
                                              style: .plain,
                                              target: self,
                                              action: nil)
        emptyButtonItem.tintColor = .clear
        emptyButtonItem.isEnabled = false
        navigationItem.leftBarButtonItems = [exitButtonItem, emptyButtonItem]
        
        let usersButtonItem = UIBarButtonItem(image: UIImage(named: "add"),
                                              style: .plain,
                                              target: self,
                                              action: #selector(didTapNewChat(_:)))
        navigationItem.rightBarButtonItem = usersButtonItem
        usersButtonItem.tintColor = .white
        addInfoButton()
    }
    
    override func setupNavigationTitle() {
        self.title = DialogsConstant.chats
    }
    
    override func setupDialogs() {
        chatManager.updateStorage()
        self.refreshControl?.beginRefreshing()
    }
    
    @objc private func didTapEdit(_ gestureReconizer: UILongPressGestureRecognizer) {
        guard gestureReconizer.state != .ended else {
            return
        }
        gestureReconizer.state = .ended
        guard let deleteVC = Screen.dialogsSelectionViewController() else {
            return
        }
        deleteVC.action = .delete
        navigationController?.pushViewController(deleteVC, animated: false)
    }
    
    @objc private func didTapNewChat(_ sender: UIBarButtonItem) {
        guard let createNewDialogVC = Screen.createNewDialogViewController() else {
            return
        }
        navigationController?.pushViewController(createNewDialogVC, animated: true)
    }
    
    @objc private func didTapLogout() {
        logout()
    }
    
    private func logout() {
        if connection.established == false {
            showAnimatedAlertView(nil, message: ConnectionConstant.connectingState)
            return
        }
        progressView.start()
        NotificationsProvider.clearSubscription { [weak self] in
            self?.connection.breakConnection {
                guard let self = self else {
                    return
                }
                self.authModule.logout()
            }
        }
    }
    
    @IBAction func refreshDialogs(_ sender: UIRefreshControl) {
        chatManager.updateStorage()
    }
    
    //MARK: - Internal Methods
    private func openChatScreen(_ dialogID: String) {
        guard let chatVC = Screen.chatViewController() else {
            return
        }
        chatVC.dialogID = dialogID
        navigationController?.pushViewController(chatVC, animated: false)
    }
    
    // MARK: - UITableViewDataSource
    override func configure(cell: DialogCell, for indexPath: IndexPath) {
        let dialog = dialogs[indexPath.row]
        tableView.allowsMultipleSelection = false
        cell.checkBoxImageView.isHidden = true
        cell.checkBoxView.isHidden = true
        cell.lastMessageDateLabel.isHidden = false
        cell.contentView.backgroundColor = .clear
        cell.lastMessageDateLabel.isHidden = false
        cell.unreadMessageCounterLabel.isHidden = false
        if dialog.type != .publicGroup {
            cell.unreadMessageCounterLabel.text = dialog.unreadMessagesCounter
            cell.unreadMessageCounterHolder.isHidden = dialog.unreadMessagesCounter == nil
        }
    }
    
    // MARK: - UITableViewDelegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let dialog = dialogs[indexPath.row]
        guard let dialogID = dialog.id else {
            return
        }
        openChatScreen(dialogID)
    }
    
    //MARK: - Helper Methods
    fileprivate func showSplashScreen() {
        if splashVC == nil, let splashVC = Screen.splashScreenController() {
            self.splashVC = splashVC
            splashVC.modalPresentationStyle = .overCurrentContext
            self.present(splashVC, animated: false)
        }
    }
    
    fileprivate func hideSplashScreen() {
        if splashVC == nil {
            return
        }
        splashVC?.dismiss(animated: false, completion: {
            self.splashVC = nil
        })
        progressView.stop()
    }
}

// MARK: - QBChatDelegate
extension DialogsViewController: QBChatDelegate {
    func chatRoomDidReceive(_ message: QBChatMessage, fromDialogID dialogID: String) {
        chatManager.updateDialog(with: dialogID, with: message)
    }
    
    func chatDidReceive(_ message: QBChatMessage) {
        guard let dialogID = message.dialogID else {
            return
        }
        chatManager.updateDialog(with: dialogID, with: message)
    }
    
    func chatDidReceiveSystemMessage(_ message: QBChatMessage) {
        guard let dialogID = message.dialogID else {
            return
        }
        chatManager.updateDialog(with: dialogID, with: message)
    }
}

// MARK: - ChatManagerDelegate
extension DialogsViewController: ChatManagerDelegate {
    func chatManager(_ chatManager: ChatManager, didUpdateChatDialog chatDialog: QBChatDialog) {
        reloadContent()
    }
    
    func chatManager(_ chatManager: ChatManager, didFailUpdateStorage message: String) {
        refreshControl?.endRefreshing()
        showAnimatedAlertView(nil, message: message)
    }
    
    func chatManager(_ chatManager: ChatManager, didUpdateStorage message: String) {
        debugPrint("[DialogsViewController] didUpdateStorage")
        reloadContent()
    }
}

// MARK: - AuthModuleDelegate
extension DialogsViewController: AuthModuleDelegate {
    func authModule(_ authModule: AuthModule, didLoginUser user: QBUUser) {
        Profile.synchronize(withUser: user)
        connection.establish()
        let userDefaults = UserDefaults.standard
        guard userDefaults.bool(forKey: NotificationsConstant.needUpdateToken) != false,
              let token = userDefaults.object(forKey: NotificationsConstant.token) as? Data else {
            return
        }
        NotificationsProvider.deleteLastSubscription {
            NotificationsProvider.createSubscription(withToken: token)
        }
    }
    
    func authModuleDidLogout(_ authModule: AuthModule) {
        connection.deactivateAutomaticMode()
        navigationController?.popToRootViewController(animated: false)
        Profile.clear()
        chatManager.storage.clear()
        CacheManager.shared.clearCache()
        progressView.stop()
        onSignOut?()
    }
    
    func authModule(_ authModule: AuthModule, didReceivedError error: ErrorInfo) {
        showUnAuthorizeAlert(message: error.info, logoutAction: { [weak self] action in
            self?.logout()
        }, tryAgainAction: { action in
            let profile = Profile()
            authModule.login(fullName: profile.fullName, login: profile.login)
        })
    }
}

// MARK: - ConnectionModuleDelegate
extension DialogsViewController: ConnectionModuleDelegate {
    func connectionModuleWillConnect(_ connectionModule: ConnectionModule) {
        showAnimatedAlertView(nil, message: ConnectionConstant.connectingState)
    }
    
    func connectionModuleDidConnect(_ connectionModule: ConnectionModule) {
        setupDialogs()
        hideAlertView()
        hideSplashScreen()
    }
    
    func connectionModuleDidNotConnect(_ connectionModule: ConnectionModule, error: Error) {
        hideSplashScreen()
        refreshControl?.endRefreshing()
        if error._code.isNetworkError {
            showNoInternetAlert(handler: nil)
            return
        }
        showAlertView(nil, message: error.localizedDescription, handler: nil)
    }
    
    func connectionModuleWillReconnect(_ connectionModule: ConnectionModule) {
        refreshControl?.endRefreshing()
        showAnimatedAlertView(nil, message: ConnectionConstant.reconnectingState)
    }
    
    func connectionModuleDidReconnect(_ connectionModule: ConnectionModule) {
        setupDialogs()
        hideAlertView()
    }
    
    func connectionModuleTokenHasExpired(_ connectionModule: ConnectionModule) {
        showSplashScreen()
        refreshControl?.endRefreshing()
        let profile = Profile()
        authModule.login(fullName: profile.fullName, login: profile.login)
    }
}
